### PR TITLE
Spontaneous extra day

### PR DESCRIPTION
**Pull request by rbsPREET**
* [Trello link](https://trello.com/c/OyF6HSrg/540-what-a-spontaneity-pop-up-change-to-pop-a-day-before)
* Changes to UpcomingDonationScreen.tsx file
* UI text changed instead of today to 48 hours (**please check in the local server**)
pic

### CODE REVIEW

File: packages/donor/src/screens/UpcommingDonation/UpcomingDonationScreen.tsx [hunk 1]
```diff
@@ -132,17 +141,11 @@ function NeedRideButton(props: { hospital: Hospital }) {
         name="ask_for_a_ride"
         open={open}
         title="אין לך איך להגיע?"
-        buttonApproveText="בקשת הסעה"
+        buttonApproveText={buttonApproveText}
```
Comment (evbambly) [inline packages/donor/src/screens/UpcommingDonation/UpcomingDonationScreen.tsx:144]: See previous, no need to change the text

File: packages/donor/src/screens/UpcommingDonation/SameDayDonationUtil.ts [hunk 2]
```diff
@@ -5,9 +5,21 @@ export function shouldDisplaySameDayDonationPopup(
   bookedAppointment: BookedAppointment
 ) {
   const now = getNow();
+  const tomorrow = new Date(now);
+  tomorrow.setDate(tomorrow.getDate() + 1);
   const nowMillis = now.getTime();
   const donationStartTime = new Date(bookedAppointment.donationStartTimeMillis);
-  if (!DateUtils.isSameDay(now, donationStartTime)) {
+  // Checks if the dates are in the same day OR checking if it is in the last 48 hours
```
Comment (evbambly) [inline packages/donor/src/screens/UpcommingDonation/SameDayDonationUtil.ts:12]: Could this if statement be simplified?

**Review by evbambly — CHANGES_REQUESTED** [review]

File: packages/donor/src/screens/UpcommingDonation/UpcomingDonationScreen.tsx [hunk 3]
```diff
@@ -116,15 +116,14 @@ function NeedRideButton(props: { hospital: Hospital }) {
   let onApprovedContent = "";
   switch (props.hospital) {
     case Hospital.BEILINSON:
-      content = "כדי לתאם הסעה ניתן להתקשר למיכל, מתאמת התרומות בבילינסון";
-      buttonApproveText = "התקשר למתאמת התרומות";
```
Comment (evbambly) [inline packages/donor/src/screens/UpcommingDonation/UpcomingDonationScreen.tsx:120]: What is the content of buttonApproveText?

File: packages/donor/src/screens/UpcommingDonation/UpcomingDonationScreenContainer.tsx [hunk 4]
```diff
@@ -31,6 +34,7 @@ export default function UpcomingDonationScreenContainer(
     props.setBookedAppointment(undefined);
   };
 
+  /*
```
Comment (evbambly) [inline packages/donor/src/screens/UpcommingDonation/UpcomingDonationScreenContainer.tsx:37]: Why is this commented out?

File: packages/donor/src/screens/UpcommingDonation/UpcomingDonationScreen.tsx (hunk 1 above)
Comment (evbambly) [reply packages/donor/src/screens/UpcommingDonation/UpcomingDonationScreen.tsx]: Unresolved

File: packages/donor/src/screens/UpcommingDonation/TimeFrameAfterDonationUtil.tsx [hunk 5]
```diff
@@ -0,0 +1,34 @@
+import { BookedAppointment, DateUtils } from "@zm-blood-components/common";
+
+export interface ITimeFrameAfterDonationPopupProps {
```
Comment (evbambly) [inline packages/donor/src/screens/UpcommingDonation/TimeFrameAfterDonationUtil.tsx:3]: Please keep Interface naming convention consistent with the function name

File: packages/donor/src/screens/UpcommingDonation/TimeFrameAfterDonationUtil.tsx [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+  bookedAppointment: BookedAppointment;
+  daysLimit: number; // 1 = 24 hours, 2 = 48 hours, ...
+}
+
+export function TimeFrameAfterDonationPopup({
+  getCurrentDateTime,
+  bookedAppointment,
+  daysLimit,
+}: ITimeFrameAfterDonationPopupProps) {
+  const currentDayTime = getCurrentDateTime();
```
Comment (evbambly) [inline packages/donor/src/screens/UpcommingDonation/TimeFrameAfterDonationUtil.tsx:14]: Why should TimeFrameAfterDonationPopup receive getCurrentDateTime as a prop?

File: packages/donor/src/screens/UpcommingDonation/TimeFrameAfterDonationUtil.tsx [hunk 7]
```diff
@@ -0,0 +1,34 @@
+import { BookedAppointment, DateUtils } from "@zm-blood-components/common";
+
+export interface ITimeFrameAfterDonationPopupProps {
+  getCurrentDateTime: () => Date;
+  bookedAppointment: BookedAppointment;
+  daysLimit: number; // 1 = 24 hours, 2 = 48 hours, ...
+}
+
+export function TimeFrameAfterDonationPopup({
```
Comment (evbambly) [inline packages/donor/src/screens/UpcommingDonation/TimeFrameAfterDonationUtil.tsx:9]: Could we keep the shouldDisplayPopup naming format?

File: packages/donor/src/screens/UpcommingDonation/TimeFrameAfterDonationUtil.tsx [hunk 8]
```diff
@@ -0,0 +1,34 @@
+import { BookedAppointment, DateUtils } from "@zm-blood-components/common";
+
+export interface ITimeFrameAfterDonationPopupProps {
+  getCurrentDateTime: () => Date;
+  bookedAppointment: BookedAppointment;
+  daysLimit: number; // 1 = 24 hours, 2 = 48 hours, ...
```
Comment (evbambly) [inline packages/donor/src/screens/UpcommingDonation/TimeFrameAfterDonationUtil.tsx:6]: Which component should be responsible for this number?

File: packages/donor/src/screens/UpcommingDonation/TimeFrameAfterDonationUtil.tsx [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+  daysLimit,
+}: ITimeFrameAfterDonationPopupProps) {
+  const currentDayTime = getCurrentDateTime();
+  const numberOfDaysBetweenDates = DateUtils.getNumberOfDaysBetweenDates(
+    bookedAppointment.donationStartTimeMillis,
+    currentDayTime.getTime()
+  );
+  const currentDayMillis = currentDayTime.getTime();
+  // Checks if the donation is in the time frame
+  if (!(numberOfDaysBetweenDates <= daysLimit)) {
```
Comment (evbambly) [inline packages/donor/src/screens/UpcommingDonation/TimeFrameAfterDonationUtil.tsx:21]: Could we make this expression clearer?

**Review by evbambly — CHANGES_REQUESTED** [review]

File: packages/donor/src/screens/UpcommingDonation/TimeFrameAfterDonationUtil.tsx (hunk 7 above)
Comment (evbambly) [reply packages/donor/src/screens/UpcommingDonation/TimeFrameAfterDonationUtil.tsx:9]: Can we find a more generic name?

File: packages/donor/src/screens/UpcommingDonation/TimeFrameAfterDonationUtil.tsx (hunk 8 above)
Comment (evbambly) [reply packages/donor/src/screens/UpcommingDonation/TimeFrameAfterDonationUtil.tsx:6]: DAYS_LIMIT

File: packages/donor/src/screens/UpcommingDonation/TimeFrameAfterDonationUtil.tsx (hunk 7 above)
Comment (evbambly) [reply packages/donor/src/screens/UpcommingDonation/TimeFrameAfterDonationUtil.tsx:9]: IsAppointmentClose
Please move to AppointmentUtils as per out convo.

File: packages/donor/src/screens/UpcommingDonation/TimeFrameAfterDonationUtil.test.ts [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+    id: "id",
+    donationStartTimeMillis: tempDate,
+    bookingTimeMillis: tempDate,
+    donorId: "donorId",
+    status: AppointmentStatus.BOOKED,
+    hospital: Hospital.TEL_HASHOMER,
+    booked: true,
+  } as BookedAppointment;
+
+  it("Should not show popup when appointment is more than 2 days (48 hours)", () => {
```
Comment (evbambly) [inline packages/donor/src/screens/UpcommingDonation/TimeFrameAfterDonationUtil.test.ts:23]: Should not show popup when appointment is booked for more than 2 days (48 hours) in the future

File: packages/donor/src/screens/UpcommingDonation/TimeFrameAfterDonationUtil.tsx (hunk 9 above)
Comment (evbambly) [reply packages/donor/src/screens/UpcommingDonation/TimeFrameAfterDonationUtil.tsx:21]: (numberOfDaysBetweenDates > daysLimit || numberOfDaysBetweenDates  < 0) {
return false
}

File: packages/common/src/AppointmentUtils.ts [hunk 11]
```diff
@@ -1,5 +1,32 @@
-import { MANUAL_DONOR_ID } from "./types";
+import { MANUAL_DONOR_ID, BookedAppointment } from "./types";
+import { DateUtils } from "./index";
 
 export function isManualDonor(donorId: string) {
   return donorId === MANUAL_DONOR_ID;
 }
+
+export function isAppointmentClosed({
```
Comment (evbambly) [inline packages/common/src/AppointmentUtils.ts:8]: isAppointmentClose without 'd' :)

Comment (evbambly) [reply packages/common/src/AppointmentUtils.ts:8]: Also, no need to deconstruct bookedAppointment and reconstruct it immediately.

File: packages/common/src/AppointmentUtils.ts [hunk 12]
```diff
@@ -1,5 +1,32 @@
...
+  ...bookedAppointment
+}: BookedAppointment) {
+  const DAYS_LIMIT = 2;
```
Comment (evbambly) [inline packages/common/src/AppointmentUtils.ts:11]: Maybe this should be outside the function?
And please add a clarifying comment about this variable.

Comment (rbsPREET) [reply packages/common/src/AppointmentUtils.ts:11]: Since it is a Utils folder and there is more functions in it, keeping it inside seems better

Comment (evbambly) [reply packages/common/src/AppointmentUtils.ts]: Please see implementation in 
HospitalUtils.ts #14
DateUtils.ts #15

**Review by evbambly — APPROVED** [review]

**Pull request merged**